File: ios/ImageScanner.swift
import Foundation
import React
import MLKitBarcodeScanning
import MLKitVision

@objc(ImageScanner)
class ImageScanner: NSObject {
    private var scanner: BarcodeScanner = BarcodeScanner.barcodeScanner()
    private var scannerBuilder: BarcodeScannerOptions = BarcodeScannerOptions(formats: .all)

    @objc(process:options:withResolver:withRejecter:)
    private func process(
        uri: String,
        options: [AnyHashable: Any]! = [:],
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        let scannerBarcodeFormats = ScannerUtils.getOptionsBarcodeFormats(options: options)
        let scannerRatio = ScannerUtils.getOptionsRatio(options: options)
        let scannerOrientation = ScannerUtils.getOptionsOrientation(options: options)
        let scannerViewSize = ScannerUtils.getOptionsViewSize(options: options)

        let barcodeFormats = ScannerUtils.getSafeBarcodeFormats(formats: scannerBarcodeFormats)
        if barcodeFormats.contains(.all) {
            scannerBuilder = BarcodeScannerOptions(formats: .all)
        } else {
            scannerBuilder = BarcodeScannerOptions(formats: BarcodeFormat(barcodeFormats))
        }

        scanner = BarcodeScanner.barcodeScanner(options: scannerBuilder)

        var array: [Any] = []

        guard let imageUI = UIImage(contentsOfFile: uri) else {
            reject("Error", "Can't find photo.", nil)
            return
        }

        guard let imageCG = imageUI.cgImage else {
            reject("Error", "Couldn't get CGImage from UIImage.", nil)
            return
        }

        let image = VisionImage(image: imageUI)
        image.orientation = imageUI.imageOrientation
        let imageSize = Size(width: imageCG.width, height: imageCG.height)
        let viewSize = ScannerUtils.getSafeViewSize(imageSize: imageSize, viewSize: scannerViewSize)

        do {
            let barcodes = try scanner.results(in: image)
            let barcodesFiltered: [Barcode]
            if scannerRatio.width != 1.0 || scannerRatio.height != 1.0 {
                barcodesFiltered = ScannerUtils.filterBarcodes(
                    barcodes: barcodes,
                    imageSize: imageSize,
                    viewSize: viewSize,
                    ratio: scannerRatio,
                    imageRotation: image.orientation
                )
            } else {
                barcodesFiltered = barcodes
            }

            for barcode in barcodesFiltered {
                let map = ScannerUtils.formatBarcode(
                    barcode: barcode,
                    imageSize: imageSize,
                    viewSize: viewSize,
                    orientation: scannerOrientation,
                    imageRotation: image.orientation
                )
                array.append(map)
            }
            resolve(array)
        } catch {
            reject("Error", "Image processing failed.", nil)
        }
    }
}

File: android/src/main/java/com/visioncamerabarcodesscanner/ImageScanner.kt
package com.visioncamerabarcodesscanner

import android.net.Uri

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableNativeArray

import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.barcode.BarcodeScannerOptions
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.barcode.common.Barcode
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_ALL_FORMATS

class ImageScanner(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    companion object {
        const val NAME = "ImageScanner"
    }

    private var scanner = BarcodeScanning.getClient()
    private val scannerBuilder = BarcodeScannerOptions.Builder()

    override fun getName() = NAME

    @ReactMethod
    fun process(uri: String, options: MutableMap<String, Any>?, promise: Promise) {
        val scannerBarcodeFormats = ScannerUtils.getOptionsBarcodeFormats(options)
        val scannerRatio = ScannerUtils.getOptionsRatio(options)
        val scannerOrientation = ScannerUtils.getOptionsOrientation(options)
        val scannerViewSize = ScannerUtils.getOptionsViewSize(options)

        val barcodeFormats = ScannerUtils.getSafeBarcodeFormats(scannerBarcodeFormats)
        if (barcodeFormats.contains(FORMAT_ALL_FORMATS)) {
            scannerBuilder.setBarcodeFormats(FORMAT_ALL_FORMATS)
        } else {
            val firstFormat = barcodeFormats.first()
            val otherFormats = barcodeFormats.drop(1).toIntArray()
            scannerBuilder.setBarcodeFormats(firstFormat, *otherFormats)
        }

        scanner = BarcodeScanning.getClient(scannerBuilder.build())

        val array = WritableNativeArray()

        val imageUri = Uri.parse(uri)
        // Loads image and applies EXIF rotation. width/height are already corrected.
        val image = InputImage.fromFilePath(this.reactApplicationContext, imageUri)
        val imageSize = Size(image.width, image.height)
        val viewSize = ScannerUtils.getSafeViewSize(imageSize, scannerViewSize)

        val task: Task<List<Barcode>> = scanner.process(image)
        val barcodes: List<Barcode> = Tasks.await(task)

        // Filter barcodes if scanning ratio is restricted.
        val barcodesFiltered =
            if (scannerRatio.width != 1f || scannerRatio.height != 1f)
                ScannerUtils.filterBarcodes(barcodes, imageSize, viewSize, scannerRatio)
            else
                barcodes

        for (barcode in barcodesFiltered) {
            val map = ScannerUtils.formatBarcode(barcode, imageSize, viewSize, scannerOrientation)
            array.pushMap(map)
        }
        promise.resolve(array)
    }
}
